File: frontend/lib/core/models/access.dart
import 'package:equatable/equatable.dart';
import 'user.dart';

enum AccessType { resident, visitor, service, delivery, emergency }
enum AccessStatus { pending, approved, rejected, expired, completed }

class AccessLog extends Equatable {
  final String id;
  final String? unitId;
  final UnitInfo? unit;
  final String? residentId;
  final ResidentInfo? resident;
  final String? visitorId;
  final VisitorInfo? visitor;
  final AccessType type;
  final AccessStatus status;
  final DateTime? entryTime;
  final DateTime? exitTime;
  final DateTime? scheduledEntry;
  final DateTime? scheduledExit;
  final String? entryMethod;
  final String? exitMethod;
  final String? plateRecognized;
  final bool faceRecognized;
  final String? notes;
  final DateTime createdAt;

  const AccessLog({
    required this.id,
    this.unitId,
    this.unit,
    this.residentId,
    this.resident,
    this.visitorId,
    this.visitor,
    required this.type,
    required this.status,
    this.entryTime,
    this.exitTime,
    this.scheduledEntry,
    this.scheduledExit,
    this.entryMethod,
    this.exitMethod,
    this.plateRecognized,
    this.faceRecognized = false,
    this.notes,
    required this.createdAt,
  });

  String get displayName {
    if (resident != null) return resident!.fullName;
    if (visitor != null) return visitor!.fullName;
    return 'Desconocido';
  }

  String get unitDisplay => unit != null ? '${unit!.block ?? ''}${unit!.number}' : 'N/A';

  factory AccessLog.fromJson(Map<String, dynamic> json) => AccessLog(
    id: json['id'],
    unitId: json['unitId'],
    unit: json['unit'] != null ? UnitInfo.fromJson(json['unit']) : null,
    residentId: json['residentId'],
    resident: json['resident'] != null ? ResidentInfo.fromJson(json['resident']) : null,
    visitorId: json['visitorId'],
    visitor: json['visitor'] != null ? VisitorInfo.fromJson(json['visitor']) : null,
    type: AccessType.values.firstWhere((e) => e.name.toLowerCase() == json['type'].toLowerCase()),
    status: AccessStatus.values.firstWhere((e) => e.name.toLowerCase() == json['status'].toLowerCase()),
    entryTime: json['entryTime'] != null ? DateTime.parse(json['entryTime']) : null,
    exitTime: json['exitTime'] != null ? DateTime.parse(json['exitTime']) : null,
    scheduledEntry: json['scheduledEntry'] != null ? DateTime.parse(json['scheduledEntry']) : null,
    scheduledExit: json['scheduledExit'] != null ? DateTime.parse(json['scheduledExit']) : null,
    entryMethod: json['entryMethod'],
    exitMethod: json['exitMethod'],
    plateRecognized: json['plateRecognized'],
    faceRecognized: json['faceRecognized'] ?? false,
    notes: json['notes'],
    createdAt: DateTime.parse(json['createdAt']),
  );

  @override
  List<Object?> get props => [id, type, status, entryTime];
}

class ResidentInfo extends Equatable {
  final String id;
  final String firstName;
  final String lastName;

  const ResidentInfo({required this.id, required this.firstName, required this.lastName});
  String get fullName => '$firstName $lastName';

  factory ResidentInfo.fromJson(Map<String, dynamic> json) => ResidentInfo(
    id: json['id'] ?? json['user']?['id'],
    firstName: json['user']?['firstName'] ?? json['firstName'] ?? '',
    lastName: json['user']?['lastName'] ?? json['lastName'] ?? '',
  );

  @override
  List<Object?> get props => [id, firstName, lastName];
}

class VisitorInfo extends Equatable {
  final String id;
  final String firstName;
  final String lastName;

  const VisitorInfo({required this.id, required this.firstName, required this.lastName});
  String get fullName => '$firstName $lastName';

  factory VisitorInfo.fromJson(Map<String, dynamic> json) => VisitorInfo(
    id: json['id'],
    firstName: json['firstName'] ?? '',
    lastName: json['lastName'] ?? '',
  );

  @override
  List<Object?> get props => [id, firstName, lastName];
}

File: backend/src/routes/access.ts
import { Router } from 'express';
import { prisma } from '../index.js';
import { asyncHandler, AppError } from '../middleware/errorHandler.js';
import { AuthRequest } from '../middleware/auth.js';
import {
  createAccessLogSchema,
  approveAccessSchema,
  paginationSchema,
  dateRangeSchema,
} from '../validators/schemas.js';

const router = Router();

router.get('/', asyncHandler(async (req: AuthRequest, res) => {
  const { page, limit, sortBy, sortOrder } = paginationSchema.parse(req.query);
  const { startDate, endDate } = dateRangeSchema.parse(req.query);
  const { type, status, unitId, residentId } = req.query;
  
  const where: any = { complexId: req.user!.complexId };
  if (type) where.type = type;
  if (status) where.status = status;
  if (unitId) where.unitId = unitId;
  if (residentId) where.residentId = residentId;
  if (startDate || endDate) {
    where.entryTime = {};
    if (startDate) where.entryTime.gte = new Date(startDate);
    if (endDate) where.entryTime.lte = new Date(endDate);
  }

  const [logs, total] = await Promise.all([
    prisma.accessLog.findMany({
      where,
      include: {
        unit: { select: { number: true, block: true } },
        resident: { select: { user: { select: { firstName: true, lastName: true } } } },
        visitor: { select: { firstName: true, lastName: true } },
        securityUser: { select: { firstName: true, lastName: true } },
      },
      orderBy: { [sortBy || 'entryTime']: sortOrder },
      skip: (page - 1) * limit,
      take: limit,
    }),
    prisma.accessLog.count({ where }),
  ]);

  res.json({ data: logs, pagination: { page, limit, total, totalPages: Math.ceil(total / limit) } });
}));

router.get('/active', asyncHandler(async (req: AuthRequest, res) => {
  const logs = await prisma.accessLog.findMany({
    where: {
      complexId: req.user!.complexId,
      status: 'APPROVED',
      exitTime: null,
    },
    include: {
      unit: { select: { number: true, block: true } },
      resident: { select: { user: { select: { firstName: true, lastName: true } } } },
      visitor: { select: { firstName: true, lastName: true } },
    },
    orderBy: { entryTime: 'asc' },
  });

  res.json(logs);
}));

router.get('/:id', asyncHandler(async (req: AuthRequest, res) => {
  const log = await prisma.accessLog.findFirst({
    where: { id: req.params.id, complexId: req.user!.complexId },
    include: {
      unit: { select: { number: true, block: true } },
      resident: { select: { user: { select: { firstName: true, lastName: true } } } },
      visitor: { select: { firstName: true, lastName: true } },
      securityUser: { select: { firstName: true, lastName: true } },
    },
  });

  if (!log) throw new AppError(404, 'Registro de acceso no encontrado');
  res.json(log);
}));

router.post('/', asyncHandler(async (req: AuthRequest, res) => {
  const data = createAccessLogSchema.parse(req.body);
  
  let residentId: string | undefined;
  if (data.type === 'RESIDENT' || data.type === 'VISITOR') {
    if (req.user!.role === 'RESIDENT') {
      const resident = await prisma.resident.findUnique({ where: { userId: req.user!.id } });
      if (!resident) throw new AppError(404, 'Perfil de residente no encontrado');
      residentId = resident.id;
    } else {
      if (!data.unitId && !data.visitorId) throw new AppError(400, 'unitId o visitorId requerido');
      if (data.unitId) {
        const unit = await prisma.unit.findFirst({ where: { id: data.unitId, complexId: req.user!.complexId } });
        if (!unit) throw new AppError(404, 'Unidad no encontrada');
        const resident = await prisma.resident.findFirst({ where: { unitId: data.unitId, status: 'ACTIVE' } });
        if (resident) residentId = resident.id;
      }
    }
  }

  const log = await prisma.accessLog.create({
    data: {
      ...data,
      complexId: req.user!.complexId!,
      residentId,
      scheduledEntry: data.scheduledEntry ? new Date(data.scheduledEntry) : null,
      scheduledExit: data.scheduledExit ? new Date(data.scheduledExit) : null,
      securityUserId: req.user!.role === 'SECURITY' ? req.user!.id : undefined,
    },
    include: {
      unit: { select: { number: true, block: true } },
      resident: { select: { user: { select: { firstName: true, lastName: true } } } },
      visitor: { select: { firstName: true, lastName: true } },
    },
  });

  res.status(201).json(log);
}));

router.post('/resident-entry', asyncHandler(async (req: AuthRequest, res) => {
  if (req.user!.role !== 'RESIDENT') throw new AppError(403, 'Solo residentes');
  
  const resident = await prisma.resident.findUnique({ where: { userId: req.user!.id } });
  if (!resident) throw new AppError(404, 'Perfil de residente no encontrado');

  const log = await prisma.accessLog.create({
    data: {
      complexId: req.user!.complexId!,
      unitId: resident.unitId,
      residentId: resident.id,
      type: 'RESIDENT',
      status: 'APPROVED',
      entryTime: new Date(),
      entryMethod: 'APP',
    },
  });

  res.status(201).json(log);
}));

router.post('/resident-exit', asyncHandler(async (req: AuthRequest, res) => {
  if (req.user!.role !== 'RESIDENT') throw new AppError(403, 'Solo residentes');
  
  const resident = await prisma.resident.findUnique({ where: { userId: req.user!.id } });
  if (!resident) throw new AppError(404, 'Perfil de residente no encontrado');

  const activeLog = await prisma.accessLog.findFirst({
    where: { residentId: resident.id, status: 'APPROVED', exitTime: null },
    orderBy: { entryTime: 'desc' },
  });

  if (!activeLog) throw new AppError(404, 'No hay entrada activa');

  const updated = await prisma.accessLog.update({
    where: { id: activeLog.id },
    data: { exitTime: new Date(), exitMethod: 'APP', status: 'COMPLETED' },
  });

  res.json(updated);
}));

router.post('/peatonal', asyncHandler(async (req: AuthRequest, res) => {
  if (req.user!.role !== 'RESIDENT') throw new AppError(403, 'Solo residentes');

  const resident = await prisma.resident.findUnique({ where: { userId: req.user!.id } });
  if (!resident) throw new AppError(404, 'Perfil de residente no encontrado');

  const log = await prisma.accessLog.create({
    data: {
      complexId: req.user!.complexId!,
      unitId: resident.unitId,
      residentId: resident.id,
      type: 'RESIDENT',
      status: 'APPROVED',
      entryTime: new Date(),
      entryMethod: 'PEATONAL',
    },
  });

  res.status(201).json(log);
}));

router.post('/botonera', asyncHandler(async (req: AuthRequest, res) => {
  if (req.user!.role !== 'RESIDENT') throw new AppError(403, 'Solo residentes');

  const resident = await prisma.resident.findUnique({ where: { userId: req.user!.id } });
  if (!resident) throw new AppError(404, 'Perfil de residente no encontrado');

  res.json({ ok: true, message: 'Botonera activada' });
}));

router.patch('/:id/approve', asyncHandler(async (req: AuthRequest, res) => {
  if (!['ADMIN', 'SECURITY', 'COMMITTEE'].includes(req.user!.role)) throw new AppError(403, 'Sin permisos');
  
  const data = approveAccessSchema.parse(req.body);
  
  const log = await prisma.accessLog.findFirst({
    where: { id: req.params.id, complexId: req.user!.complexId },
  });

  if (!log) throw new AppError(404, 'Registro no encontrado');

  const updateData: any = { status: data.status, notes: data.notes };
  if (data.status === 'APPROVED' && !log.entryTime) updateData.entryTime = new Date();

  const updated = await prisma.accessLog.update({
    where: { id: log.id },
    data: updateData,
    include: {
      unit: { select: { number: true, block: true } },
      resident: { select: { user: { select: { firstName: true, lastName: true } } } },
      visitor: { select: { firstName: true, lastName: true } },
    },
  });

  res.json(updated);
}));

router.patch('/:id/exit', asyncHandler(async (req: AuthRequest, res) => {
  if (!['ADMIN', 'SECURITY', 'COMMITTEE'].includes(req.user!.role)) throw new AppError(403, 'Sin permisos');
  
  const log = await prisma.accessLog.findFirst({
    where: { id: req.params.id, complexId: req.user!.complexId },
  });

  if (!log) throw new AppError(404, 'Registro no encontrado');
  if (log.exitTime) throw new AppError(400, 'Ya tiene hora de salida');

  const updated = await prisma.accessLog.update({
    where: { id: log.id },
    data: { exitTime: new Date(), exitMethod: 'MANUAL', status: 'COMPLETED' },
  });

  res.json(updated);
}));

router.get('/stats/summary', asyncHandler(async (req: AuthRequest, res) => {
  const { startDate, endDate } = dateRangeSchema.parse(req.query);
  const complexId = req.user!.complexId!;
  
  const where: any = { complexId };
  if (startDate || endDate) {
    where.entryTime = {};
    if (startDate) where.entryTime.gte = new Date(startDate);
    if (endDate) where.entryTime.lte = new Date(endDate);
  }

  const [total, residents, visitors, services, deliveries, active] = await Promise.all([
    prisma.accessLog.count({ where }),
    prisma.accessLog.count({ where: { ...where, type: 'RESIDENT' } }),
    prisma.accessLog.count({ where: { ...where, type: 'VISITOR' } }),
    prisma.accessLog.count({ where: { ...where, type: 'SERVICE' } }),
    prisma.accessLog.count({ where: { ...where, type: 'DELIVERY' } }),
    prisma.accessLog.count({ where: { complexId, status: 'APPROVED', exitTime: null } }),
  ]);

  res.json({ total, residents, visitors, services, deliveries, active });
}));

export default router;